apps/extension/src/background/keyring/keyring.ts

```diff
@@ -1,68 +1,217 @@
 import { KVStore } from "@anoma/storage";
+import { AEAD, Bip44, Mnemonic, PhraseSize } from "@anoma/crypto";
+import { Address } from "@anoma/shared";
+import { KeyRingStatus, MnemonicState, AccountState } from "./types";
+import { v5 as uuid } from "uuid";
+import { Bip44Path, DerivedAccount } from "./types";
 
-export enum KeyRingStatus {
-  NOTLOADED,
-  EMPTY,
-  LOCKED,
-  UNLOCKED,
+import { IStore, Store } from "../types";
+
+// Generated UUID namespace for uuid v5
+const UUID_NAMESPACE = "9bfceade-37fe-11ed-acc0-a3da3461b38c";
+
+// Construct unique uuid, passing in an arbitray number of arguments.
+// This could be a unique parameter of the object receiving the id,
+// or an index based on the number of existing objects in a hierarchy.
+const getId = (name: string, ...args: (number | string)[]): string => {
+  // Ensure a unique UUID
+  let uuidName = name;
+
+  // Concatenate any number of args onto the name parameter
+  args.forEach((arg) => {
+    uuidName = `${uuidName}::${arg}`;
+  });
+
+  return uuid(uuidName, UUID_NAMESPACE);
+};
+
+enum KeyRingStore {
+  Mnemonic = "mnemonic",
+  Account = "account",
 }
 
-/*
- Keyring stores keys in persistent backround.
+/**
+ * Keyring stores keys in persisted backround.
  */
 export class KeyRing {
-  private _loaded: boolean;
+  private _mnemonicStore: IStore<MnemonicState>;
+  private _accountStore: IStore<AccountState>;
+  private _password: string | undefined;
+  private _status: KeyRingStatus = KeyRingStatus.EMPTY;
 
-  private _mnemonic?: Uint8Array;
-  private _keyStore: KVStore | null;
-  private _password: string = "";
-
-  constructor() {
-    this._loaded = false;
-    this._keyStore = null;
+  constructor(kvStore: KVStore) {
+    this._mnemonicStore = new Store(KeyRingStore.Mnemonic, kvStore);
+    this._accountStore = new Store(KeyRingStore.Account, kvStore);
   }
 
   public isLocked(): boolean {
-    return this.mnemonic == null || this._password == null;
+    return !this._password;
```

Comment:
Likely we would never allow a password which is an empty string`""`, but that is also falsy. It feels as if `this._password === undefined` would be more correct here (and on line 136) as we indicate the lack of a password with `undefined`?